### ISSUE TITLE
DB Encoding Issue on Ubunutu 20.04

### ISSUE THREAD

**Issue by rjemanuele** [treydock/puppet-module-keycloak#284]
When Installing Keycloak using the the 9.4.0 module version (and the 9.3x versions):

```
Error: Error executing SQL; psql returned pid 3864291 exit 1: 'ERROR:  encoding "UTF8" does not match locale "en_US"
DETAIL:  The chosen LC_CTYPE setting requires encoding "LATIN1".
'
Error: /Stage[main]/Keycloak::Db::Postgres/Postgresql::Server::Db[keycloak]/Postgresql::Server::Database[keycloak]/Postgresql_psql[CREATE DATABASE "keycloak"]/command: change from 'notrun' to 'CREATE DATABASE "keycloak" WITH TEMPLATE = "template0" ENCODING = \'UTF8\'  ' failed: Error executing SQL; psql returned pid 3864291 exit 1: 'ERROR:  encoding "UTF8" does not match locale "en_US"
DETAIL:  The chosen LC_CTYPE setting requires encoding "LATIN1".
'
```

Either Keycloak or Postgres wants to use the incorrect encoding.

**Comment by rjemanuele** [comment]
The system has:
```
LANGUAGE=en_US:
LANG=en_US
```

Is is ubuntu 20.04.5 running Postgres 12.14.

Looks OK unless Postgres has a weird locale set.

**Comment by rjemanuele** [comment]
In `/etc/postgresql/12/main/postgresql.conf`:
```
# These settings are initialized by initdb, but they can be changed.
lc_messages = 'en_US'                   # locale for system error message
                                        # strings
lc_monetary = 'en_US'                   # locale for monetary formatting
lc_numeric = 'en_US'                    # locale for number formatting
lc_time = 'en_US'                               # locale for time formatting
```

**Comment by rjemanuele** [comment]
Here's the bit of puppet I'm using:

```
class profile::keycloak::master (
  Stdlib::Host $hostname = $title,
  String $certname = $clientcert
){
  include postgresql::server
  class { 'keycloak':
    hostname        => 'keycloak.sea1.zzzzz.com',
    http_host       => '127.0.0.1',
    proxy           => 'edge',
    db              => 'postgres',
    db_url_host     => 'localhost',
    db_url_port     => 5432,
    db_url_database => 'keycloak',
    db_username     => 'keycloak',
    db_password     => lookup('keycloak_db_password'),
    admin_user_password => lookup('keycloak_initial_password'),
  }

  keycloak_realm { lookup('company_realm'):
    ensure                   => 'present',
    display_name             => lookup('company'),
    remember_me              => true,
    login_with_email_allowed => true,
    smtp_server_host         => lookup('local_smtp_address'),
    smtp_server_starttls     => true,
    smtp_server_from         => 'keycloak-noreply@zzzzz.com',
    verify_email             => true,
    browser_flow             => "browser-webauthn"
  }

  #keycloak_flow { "browser-webauthn on ${lookup('company_realm')}":
  #  ensure => 'present'
  #}

  # Probaby want to add the flow in puppet form here
  # it does require webauthn settings but they may not be accessiby from puppet

  include apache
  apache::vhost { $hostname:
    servername => $hostname,
    port        => '443',
    ssl         => true,
    ssl_reload_on_change => false,
    suphp_engine => 'off',
    manage_docroot  => false,
    docroot         => '/var/www/html',
    proxy_preserve_host => true,
    proxy_add_headers => true,
    proxy_pass          => [
      {'path' => '/', 'url' => 'http://localhost:8080/'}
    ],
    request_headers     => [
      'set X-Forwarded-Proto "https"',
      'set X-Forwarded-Port "443"'
    ],
    ssl_cert            => "/etc/step_acme_certs/${certname}.crt",
    ssl_key             => "/etc/step_acme_certs/${certname}.key",
    ssl_chain           => "/etc/step_acme_certs/${certname}.chain.pem",
  }

}

```

**Comment by rjemanuele** [comment]
Any further thoughts on this?

**Comment by treydock** [comment]
I am not sure why Postgres really insists on wanting LATIN1, I'm not really familiar with Postgres.  One possible thing you can try is this:

```
Postgresql::Server::Db <| title == $keycloak::db_url_database |> {
  encoding => 'LATIN1',
}
```

You'd want to add this in your profile somewhere after the `keycloak` class is defined.  If that solves your problem then I can update this module to make the encoding a parameter that you can change rather than being hardcoded.  I'm not entirely sure why this would be necessary but it would at least get rid of the errors you are seeing I hope.

**Comment by treydock** [comment]
The fix will be now to use `db_encoding` parameter rather than previous work around.  This is released with v10.1.0 once Github Actions complete
